plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+{
+    if (_reply == NULL)
+        return;
+
+    redisReply * reply = (redisReply*)_reply;
+    assertCallbackError(context, reply, "callback fail");
+
+    if (reply->type == REDIS_REPLY_ARRAY && strcmp("message", reply->element[0]->str) == 0 )
+    {
+        ((StringAttr*)privdata)->set(reply->element[2]->str);
+        const char * channel = reply->element[1]->str;
+        redisAsyncCommand(context, NULL, NULL, "UNSUBSCRIBE %b", channel, strlen(channel));
+        redisLibevDelRead((void*)context->ev.data);
+    }
+}
+void AsyncConnection::unsubCB(redisAsyncContext * context, void * _reply, void * privdata)
+{
+    redisReply * reply = (redisReply*)_reply;
+    assertCallbackError(context, reply, "get callback fail");
+    redisLibevDelRead((void*)context->ev.data);
+}
+void AsyncConnection::getCB(redisAsyncContext * context, void * _reply, void * privdata)
+{
+    redisReply * reply = (redisReply*)_reply;
+    assertCallbackError(context, reply, "get callback fail");
+
+    ReturnValue * retVal = (ReturnValue*)privdata;
+    retVal->cpy((size_t)reply->len, reply->str);
+    redisLibevDelRead((void*)context->ev.data);
+}
+void AsyncConnection::setCB(redisAsyncContext * context, void * _reply, void * privdata)
+{
+    redisReply * reply = (redisReply*)_reply;
+    assertCallbackError(context, reply, "set callback fail");
+    redisLibevDelRead((void*)context->ev.data);
+}
+void AsyncConnection::setLockCB(redisAsyncContext * context, void * _reply, void * privdata)
+{
+    redisReply * reply = (redisReply*)_reply;
+    assertCallbackError(context, reply, "set lock callback fail");
+
+    switch(reply->type)
+    {
+    case REDIS_REPLY_STATUS :
+        *(bool*)privdata = strcmp(reply->str, "OK") == 0;
+        break;
+    case REDIS_REPLY_NIL :
+        *(bool*)privdata = false;
+    }
+    redisLibevDelRead((void*)context->ev.data);
+}
+void AsyncConnection::selectCB(redisAsyncContext * context, void * _reply, void * privdata)
+{
+    redisReply * reply = (redisReply*)_reply;
+    assertCallbackError(context, reply, "select callback fail");
+    redisLibevDelRead((void*)context->ev.data);
+}
+
+void AsyncConnection::pubCB(redisAsyncContext * context, void * _reply, void * privdata)
+{
+    redisReply * reply = (redisReply*)_reply;
+    assertCallbackError(context, reply, "get callback fail");
+    redisLibevDelRead((void*)context->ev.data);
+}
+void AsyncConnection::unsubscribe(const char * channel)
+{
+    attachLibev();
+    assertRedisErr(redisAsyncCommand(context, NULL, NULL, "UNSUBSCRIBE %b", channel, strlen(channel)), "UNSUBSCRIBE buffer write error");
+    redisAsyncHandleWrite(context);
+}
+void AsyncConnection::subscribe(const char * channel, StringAttr & value)
+{
+    attachLibev();
+    assertRedisErr(redisAsyncCommand(context, subCB, (void*)&value, "SUBSCRIBE %b", channel, strlen(channel)), "SUBSCRIBE buffer write error");
+    handleLoop(EV_DEFAULT_ EV_TIMEOUT);
```

Comment:
I think this needs more explanation.  The code I would expect to see would be

ensureEvLoopRunning()
asyncCommand()
response.wait()

where the callback either signals a semaphore, or interrupts it if there was an error.